### ISSUE TITLE
New feature: `hydra_zen.builds_decorator`

### ISSUE THREAD

**Issue by jgbos** [mit-ll-responsible-ai/hydra-zen#93]
We had a discussion on Teams about possibly adding a `builds_decorator` of the form:

```python
def builds_decorator(f):
    Conf = builds(f, populate_full_sig=True)
    f.config = Conf
    return f
```

Then we can automatically build configs for a function or class.  For example, to get the config for the following function,

```python
@builds_decorator
def foo(x, val=1):
  return x, val
```

simply access the `config` attribute,

```python
FooConfig = foo.config(val=2)
```

**Comment by rsokl** [comment]
I like this idea and have been trying to think of ways to make `builds` a bit more ergonomic along these lines. That being said, it feels a bit "risky" to actually ship this:

- IDEs will complain about accessing `.config(...)`

![image](https://user-images.githubusercontent.com/29104956/132761853-a0916a54-cc83-4572-8198-07087bf8b80b.png)

- There is the potential of colliding with objects that already have a `config` method.
- Getting the type-annotations right seem hard (maybe impossible)
- Mutating an object to have a new, hydra-zen specific method could lead to some unforeseen issues / probably isn't considered good practice
- Once we ship this, removing it later would break peoples experiments pretty badly

Maybe we could simply document this as a recipe for library authors to use, but they are ultimately responsible for opting-in/maintaining it. Like, I can definitely imagine `mushin` implementing this.

**Comment by jgbos** [comment]
Good point.  I just remember we were talking about so thought I'd document our discussion at least :)

**Comment by rsokl** [comment]
Yep! I think it is good to document our thoughts on these sorts of things. It can be also helpful for us to reference these issues if users request similar features later.

**Comment by rsokl** [comment]
@jgbos what do you think about this approach

```python
from functools import partial


def as_builds(target):
    return partial(builds, target, populate_full_signature=True)


RandomCrop = as_builds(transforms.RandomCrop)
RandomHorizontalFlip = as_builds(transforms.RandomHorizontalFlip)
ColorJitter = as_builds(transforms.ColorJitter)
RandomRotation = as_builds(transforms.RandomRotation)
ToTensor = as_builds(transforms.ToTensor)
Normalize = as_builds(transforms.Normalize)
Compose = as_builds(transforms.Compose)

Cifar10TrainTransformsConf = Compose(
    transforms=[
        RandomCrop(size=32, padding=4),
        RandomHorizontalFlip(),
        ColorJitter(brightness=0.25, contrast=0.25, saturation=0.25),
        RandomRotation(degrees=2),
        ToTensor(),
        Normalize(
            mean=[0.4914, 0.4822, 0.4465],
            std=[0.2023, 0.1994, 0.2010],
        ),
    ],
)
```

Obviously all of the `thing = as_builds(thing)` isn't super ideal. But maybe there is a clever way to deal with this (like `make_builds_namespace(transforms)` returns a namespace of all the public callables, but with `as_builds` applied to them?)

Some cool things about this: our runtime validation ends up being really useful and intuitive here:

```pycon
>>> Normalize(average=2)
TypeError: Building: Normalize ..
The following unexpected keyword argument(s) was specified for torchvision.transforms.transforms.Normalize via `builds`: average
```

**Comment by jgbos** [comment]
I guess I'm not sure how this really helps (other than the validation).  Are you saying that since a project is going to use `hydra-zen` it might as well convert the project's classes/funcs to output `builds` objects?  Or is this just done in the `configs.py`?

**Comment by rsokl** [comment]
> Or is this just done in the configs.py?

Yep, I am thinking that end-users could do something like this in `configs.py`. That way writing the configs is literally as easy as initializing the object itself.

Although it would be super slick if we could do something like:

```python
from torchvisions import transforms

#  some magic where we auto-apply `as_builds` to everything in `transforms.__all__` and then
# expose a corresponding namespace that exposes the config counterparts
cfg = hydrate(transforms)  

cfg.Normalize  # is actually partial(builds, transforms.Normalize)
```

and in this way, the library maintainers don't have to do anything at all to make it easy to configure their stuff.

**Comment by jgbos** [comment]
👍 to `hydrate(transforms)`

I like that.  I always hate that tab completion for imports would have `Normalize` and `NormalizeCfg`.

**Comment by addisonklinke** [comment]
> potential of colliding with objects that already have a `config` method

@rsokl What about a runtime check in the decorator - does that sufficiently handle the collision downside? 

```python
def builds_decorator(f):
    if hasattr(f, 'config'):
        raise RuntimeError(f'{f.__name__}.config already exists, refusing to override with decorator')
    f.config = builds(f, populate_full_signature=True)
    return f  
```

Maybe you would also need to wrap `f.config` in a read-only class so that future access (i.e. dynamically setting `config` on an initialized class instance) cannot delete or change the attribute populated by the decorator

**Comment by addisonklinke** [comment]
@rsokl @jgbos regarding `hydrate()`, have you seen the [hydra-torch](https://github.com/pytorch/hydra-torch) package by @romesco? I think you all are tackling a similar problem but with slightly different approaches, and I'm excited to see where this goes!

**Comment by rsokl** [comment]
> What about a runtime check in the decorator - does that sufficiently handle the collision downside?

@addisonklinke That runtime check looks good to me. We could also name it something like `zen_config` which is far less-likely to collide with a pre-existing attribute of the same name. I also agree that making this a read-only attribute is worth considering.

For now, I don't envision adding `builds_decorator` to `hydra-zen`; given that it is such a thin wrapper around `builds`, interested users can pretty easily implement it on their end. If you are one of this interested users, I would be delighted to get your feedback on how this works for you. If this proves to be an especially fruitful decorator (e.g. it makes simpler the process of authoring library code that can also be easily configured in a Hydra app), then I will happily add it to hydra-zen.

> regarding hydrate()

You might be interested in my crude prototype for [`hydra_zen.like`](https://github.com/mit-ll-responsible-ai/hydra-zen/pull/219). As described in the linked PR, this could potentially be encapsulate the proposed functionality of `hydrate`. The major blocker for both `like` and `hydrate` is getting the type-annotations right. Put plainly: I don't want to ship features that will create red squiggles in IDEs wherever they are used. 

> have you seen the [hydra-torch](https://github.com/pytorch/hydra-torch) package

We are familiar with hydra-torch, but thanks for pointing us to it! We had considered using it prior to our creating hydra-zen.

One issue with an approach like hydra-torch is that the notion of spinning up a similar codebase for other third-party libraries simply is too cumbersome and requires [too much repetition](https://mit-ll-responsible-ai.github.io/hydra-zen/explanation/dont_repeat_yourself.html). Additionally, for some torch objects -- especially  optimizers -- we often want to leverage [partial instantiation](https://mit-ll-responsible-ai.github.io/hydra-zen/how_to/partial_config.html), which isn't immediately possible with hydra-torch's configs.



https://github.com/mit-ll-responsible-ai/hydra-zen/pull/219#issue-1124424274

**Comment by addisonklinke** [comment]
> I would be delighted to get your feedback on how this works for you. If this proves to be an especially fruitful decorator (e.g. it makes simpler the process of authoring library code that can also be easily configured in a Hydra app), then I will happily add it to hydra-zen

I have built out my own, and so far I really like how it's working out. I've been trying to think of an appropriate term for what it enables, and I think the closest is "object oriented packaging". I've written the decorator in such a way that my config groups mirror the hierarchy of the application's Python package. This makes it very intuitive to decide where certain class definitions should live and a new user could also understand the package quickly by reading the CLI `--help`

Portions of the design have turned out to be more complex than I thought, and it would be great to discuss some more concrete details with you or any of the other Hydra-Zen authors. I'm actually presenting my tech stack at a [MLOps.Community](https://mlops.community/) meetup and will focus heavily on the configuration system I built to power an experimentation-friendly PyTorch Lightning application with Hydra-Zen. If you all are interested, please hop in their [Slack workspace](https://go.mlops.community/slack) and shoot me a DM - I'll get you added to the Google Meet invite. Anyone else reading this issue thread is also welcome to join

> You might be interested in my crude prototype for [`hydra_zen.like`](https://github.com/mit-ll-responsible-ai/hydra-zen/pull/219)

Thanks for the pointer! I'll continue on that PR thread

> One issue with an approach like hydra-torch is that the notion of spinning up a similar codebase for other third-party libraries simply is too cumbersome

I also find it a little too complex to require a new `pip install hydra-*` every time you want to use configs from a 3rd party library. Being able to call `hydrate` dynamically inside your application code would be much more flexible

**Comment by rsokl** [comment]
> I have built out my own, and so far I really like how it's working out. I've been trying to think of an appropriate term for what it enables, and I think the closest is "object oriented packaging". I've written the decorator in such a way that my config groups mirror the hierarchy of the application's Python package. This makes it very intuitive to decide where certain class definitions should live and a new user could also understand the package quickly by reading the CLI --help

This sounds great! I'd love to see what this design looks like. Perhaps you could point me to the code, or we could arrange a video call sometime so that we can chat about it. Ultimately, I am quite interested in facilitating this sort of design for library authors who want to create "Hydra-aware" code.

**Comment by addisonklinke** [comment]
Great looking forward to discussing further at the meetup!
